### DISCUSSION TITLE
Memory leak on Render.com server

### DISCUSSION THREAD

**Discussion by andychongyz** [remix-run/remix#8973] [Q&A]
Hi, I deployed a Remix app on Render.com as a node app, not serverless. The memory usage increases over time, and it seems like a memory leak. Is this normal, and if not, how can I identify the part of the code causing it? I'm new to tracing memory leaks with node.js apps. Thank you!

![image](https://github.com/remix-run/remix/assets/12697240/ae5aca14-253d-432e-afa7-215235eb5b53)

**Comment by chigia001** [comment]
Do anything happen at that CPU spike? That CPU spike also happen when your memory reduce, so my guess is nodes start their Major GC and your memory return to normal afterward. This is normal for Nodejs, but Nodejs GC can also block main thread so server might become unreponsive during that time.

**Reply by andychongyz** [reply]
Some updates, I periodically restart the app for now using a CRON task. I found an [article](https://kentcdodds.com/blog/fixing-a-memory-leak-in-a-production-node-js-app) from Kent on how to identify the memory leak, but I haven't had the time to work on it yet.

**Comment by geiszla** [comment]
Any updates on this or tips on how to handle it? We are also experiencing this issue, and when the garbage collection happens the requests to the server time out and return 500 errors.

I've checked the memory dump in Google Cloud, but I couldn't find anything that would take up considerable share of the heap space.

Here's a flame graph of the heap dump.
![image](https://github.com/user-attachments/assets/a373b289-ed69-48e9-a33f-59c8138b5462)
